MatchBranchSimplification: fix equal const bool assignments

tmiasko opened the pull request:
The match branch simplification is applied when target blocks contain
statements that are either equal or perform a const bool assignment with
different values to the same place.

Previously, when constructing new statements, only statements from a
single block had been examined. This lead to a misoptimization when
statements are equal because the assign the *same* const bool value to
the same place.

Fix the issue by examining statements from both blocks when deciding on
replacement.

Additionally:

* Copy discriminant instead of moving it since it might be necessary to use its
  value more than once.
* Optimize when switching on copy operand

Based on #75508.

r? @oli-obk  / @JulianKnodt

JulianKnodt committed:
Add regression test for matching on u8

tmiasko committed:
MatchBranchSimplification: copy discriminant instead of moving it

It might be necessary to use its value more than once.

tmiasko committed:
MatchBranchSimplification: optimize when switching on copy operand

JulianKnodt committed:
Add fix

This also explicitly checks that the types are `bool`. `try_eval_bool` also appears to just
succeed for `u8`, so this ensures that it actually is a bool before casting.

JulianKnodt commented on src/librustc_mir/transform/match_branches.rs:
```diff
@@ -78,7 +78,7 @@ impl<'tcx> MirPass<'tcx> for MatchBranchSimplification {
                         if let Some(c) = c.literal.try_eval_bool(tcx, param_env) {
                             let op = if c { BinOp::Eq } else { BinOp::Ne };
-                            *rhs = Rvalue::BinaryOp(op, Operand::Move(discr), const_cmp);
+                            *rhs = Rvalue::BinaryOp(op, Operand::Copy(discr), const_cmp);
```
Should this mirror the type that is matched in `discr`, rather than always be `Copy`?

tmiasko replied:
AFAIU move leaves undefined data behind and moving from it again would be undefined behaviour. At least in theory, if not currently in practice. Since we might need to use it more than once, using copy should be appropriate.

JulianKnodt replied:
Ah makes sense, I wonder if this should always be done if the enum is relatively large

JulianKnodt commented on src/librustc_mir/transform/match_branches.rs:
```diff
@@ -42,50 +69,67 @@ impl<'tcx> MirPass<'tcx> for MatchBranchSimplification {
+                            StatementKind::Assign(box (_, Rvalue::Use(Operand::Constant(s_c)))),
+                        ) => {
+                            let f_b = f_c.literal.try_eval_bool(tcx, param_env).unwrap();
+                            let s_b = s_c.literal.try_eval_bool(tcx, param_env).unwrap();
```
Will this panic if there are other assignment statements which are not bools?
i.e. `a = 3, b = 4`

JulianKnodt replied:
ah no nvm, this shouldn't panic because it would've matched earlier and skipped this block

tmiasko replied:
I added a comment explaining why unwrap is fine.

JulianKnodt reviewed (commented):
Ah woops yea this was something I overlooked when there are multiple assignments which do different things!

tmiasko commented on src/librustc_mir/transform/match_branches.rs:
```diff
@@ -42,50 +69,67 @@ impl<'tcx> MirPass<'tcx> for MatchBranchSimplification {
-                s
-            });
+                })
+                .collect::<Vec<_>>();
```
I think this would require something like pick3_mut. Perf run would be nice to verify how important this is, i.e., necessary to land this or not.

JulianKnodt replied:
Probably something that can be optimized out later so can be ignored for now

tmiasko replied:
I went ahead and removed the collect.

JulianKnodt commented on src/librustc_mir/transform/match_branches.rs:
```diff
@@ -48,7 +48,7 @@ impl<'tcx> MirPass<'tcx> for MatchBranchSimplification {
                     ref values,
                     ..
-                } if targets.len() == 2 && values.len() == 1 => {
+                } if targets.len() == 2 && values.len() == 1 && targets[0] != targets[1] => {
```
Just curious, is targets[0] == targets[1] ever hit? I imagine that would be a whole different opt pass

tmiasko replied:
I don't know. 

The CfgSimplifier optimizes switch with the same targets to a goto, and then collapses goto chains and merges blocks.

oli-obk replied:
we can make this condition an assert after the beta branch, then we have 12 weeks to find out if it's a problem :D

tmiasko committed:
MatchBranchSimplification: fix equal const bool assignments

The match branch simplification is applied when target blocks contain
statements that are either equal or perform a const bool assignment with
different values to the same place.

Previously, when constructing new statements, only statements from a
single block had been examined. This lead to a misoptimization when
statements are equal because the assign the *same* const bool value to
the same place.

Fix the issue by examining statements from both blocks when deciding on
replacement.

tmiasko committed:
MatchBranchSimplification: avoid intermediate vec allocation

oli-obk commented on src/librustc_index/vec.rs:
```diff
@@ -669,6 +669,17 @@ impl<I: Idx, T> IndexVec<I, T> {
+        let len = self.raw.len();
+        assert!(ai < len && bi < len && ci < len);
+        let ptr = self.raw.as_mut_ptr();
+        unsafe { (&mut *ptr.add(ai), &mut *ptr.add(bi), &mut *ptr.add(ci)) }
```
you can also write this in safe code by using `split_at_mut` twice

tmiasko replied:
I might be missing something, but it seems like it would require some non-trivial amount of code, because we don't know relative order of indices, but need to preserve the order, so there are six different cases to consider.

oli-obk replied:
ah... that's a good point indeed. I thought they were ordered

oli-obk commented:
r=me with the unsafe code replaced with a safe solution

oli-obk commented:
@bors r+

bors commented:
:pushpin: Commit af9b9e4ec874c4bc5c771841d491e02d02d5636a has been approved by `oli-obk`

<!-- @bors r=oli-obk af9b9e4ec874c4bc5c771841d491e02d02d5636a -->
<!-- homu: {"type":"Approved","sha":"af9b9e4ec874c4bc5c771841d491e02d02d5636a","approver":"oli-obk"} -->

bors commented:
:hourglass: Testing commit af9b9e4ec874c4bc5c771841d491e02d02d5636a with merge b9db927b68f4b2e9c3e648be35b412ac1839ec54...
<!-- homu: {"type":"BuildStarted","head_sha":"af9b9e4ec874c4bc5c771841d491e02d02d5636a","merge_sha":"b9db927b68f4b2e9c3e648be35b412ac1839ec54"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/988427212), [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=36007)
Approved by: oli-obk
Pushing b9db927b68f4b2e9c3e648be35b412ac1839ec54 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"oli-obk","base_ref":"master","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=36007","checks-actions":"https://github.com/rust-lang-ci/rust/runs/988427212"},"merge_sha":"b9db927b68f4b2e9c3e648be35b412ac1839ec54"} -->
